modality-probe-debug-collector/src/cli.rs

```diff
@@ -0,0 +1,666 @@
+use std::error::Error;
+use std::fmt;
+use std::fs::File;
+use std::io::prelude::*;
+use std::net::SocketAddrV4;
+use std::path::PathBuf;
+
+use parse_duration::parse;
+use structopt::StructOpt;
+
+use goblin::container::Endian;
+use goblin::elf::Elf;
+
+use modality_probe_debug_collector::{Config, ProbeAddr};
+
+/// Error in options processing
+#[derive(Debug)]
+struct OptionsError {
+    details: String,
+}
+
+impl OptionsError {
+    fn new(msg: &str) -> Self {
+        OptionsError {
+            details: msg.to_string(),
+        }
+    }
+}
+
+impl fmt::Display for OptionsError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.details)
+    }
+}
+
+impl Error for OptionsError {
+    fn description(&self) -> &str {
+        &self.details
+    }
+}
+
+#[derive(Debug, Default, StructOpt)]
+#[structopt(
+    name = "modality-probe-debug-collector",
+    about = "Periodically collects logs from microcontrollers over debug interfaces; outputs them to a file."
+)]
+pub struct CLIOptions {
+    /// Session id to associate with the collected trace data
+    #[structopt(short = "s", long = "session-id", default_value = "0")]
+    session_id: u32,
+
+    /// Specifies little endian architecture of target system
+    #[structopt(short = "l", long = "little-endian", conflicts_with = "big-endian")]
+    little_endian: bool,
+
+    /// Specifies big endian architecture of target system
+    #[structopt(short = "b", long = "big-endian")]
+    big_endian: bool,
+
+    /// Path of ELF file for endianness and symbol resolution
+    #[structopt(short = "e", long = "elf", parse(from_os_str))]
+    elf_path: Option<PathBuf>,
+
+    /// Chip type of target system to attach to
+    #[structopt(
+        short = "a",
+        long = "attach",
+        conflicts_with = "gdb-addr",
+        required_unless = "gdb-addr"
+    )]
+    attach_target: Option<String>,
+
+    /// Address of gdb server to connect to
+    #[structopt(short = "g", long = "gdb-addr", required_unless = "attach-target")]
+    gdb_addr: Option<SocketAddrV4>,
+
+    /// Interval between log collections. Ex: "2 min 15 sec 500 milli 250 micro"
+    #[structopt(short = "i", long = "interval")]
+    interval: String,
+
+    /// Output file location
+    #[structopt(short = "o", long = "output-file", parse(from_os_str))]
+    output_path: PathBuf,
+
+    /// Symbols and/or addresses of probes or probe pointers
+    probe_syms: Vec<String>,
+}
+
+/// Turn CLI options into configuration for the collector
+pub(crate) fn config_from_options(options: CLIOptions) -> Result<Config, Box<dyn Error>> {
+    let mut probe_addrs = Vec::new();
+    let mut symbols = Vec::new();
+    // Separate probe/probe pointer addresses and probe symbols
+    for addr_str in options.probe_syms.iter().filter(|s| s.starts_with("0x")) {
```

Comment:
Naive question -- do we want/need to support addresses prefixed with `0X` as well as `0x` ?